// webassembly-interpreter/lib/compiler/compile/module.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCompiledModule = createCompiledModule;
exports.Module = void 0;

var _validation = _interopRequireDefault(require("../validation"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _require = require("../AST/traverse"),
    traverse = _require.traverse;

var wastIdentifierToIndex = require("../AST/transform/wast-identifier-to-index");

var Module = function Module(ast, exports, imports) {
  _classCallCheck(this, Module);

  (0, _validation.default)(ast);
  this._ast = ast;
  this.exports = exports;
  this.imports = imports;
};

exports.Module = Module;

function createCompiledModule(ast) {
  var exports = [];
  var imports = [];
  traverse(ast, {
    ModuleExport: function (_ModuleExport) {
      function ModuleExport(_x) {
        return _ModuleExport.apply(this, arguments);
      }

      ModuleExport.toString = function () {
        return _ModuleExport.toString();
      };

      return ModuleExport;
    }(function (_ref) {
      var node = _ref.node;

      if (node.descr.type === "Func") {
        exports.push({
          name: node.name,
          kind: "function"
        });
      }
    })
  }); // Do compile-time ast manipulation in order to remove WAST
  // semantics during execution

  wastIdentifierToIndex.transform(ast);
  return new Module(ast, exports, imports);
}
